### DISCUSSION TITLE
BMC Relay Rotation

### DISCUSSION THREAD

**Discussion by quangtran8588** [icon-project/btp#8] [Q&A] [answered]
Dear folks,

I have a couple of questions relating to BMC Relay Rotation since an implementation in python, pyscore client, has not yet completed. From my understanding, BTPLink defines SACK variables (i.e. sack_term, sack_next,....) that can be used to synchronize data transferring between the Relay and BMC through BTP Message.

- When BMC receives a BTP Message for ***'svc = _sack'***, ***'sack_height'*** and ***'sack_seq'*** variables are updated. What is ***'sack_height'*** being used for thereafter?
```
elif svc == "_sack":
            sack = SACKMessage.from_bytes(msg.payload)
            link = self._links[prev]
            link.sack_height = sack.height
            link.sack_seq = sack.seq
```
- When a BMC owner registers a link, ***'sack_term'*** and ***'sack_next'*** are initialized an initial value (value = 0). Then, these two variables will be used as conditional requirements to send ***_sack_message***  when a BMC receives a BTP Message from a Relay. As requiring, ***link.sack_term*** has to be greater than zero, whereas ***link.sack_next*** has to be less than or equal to ***self.block_height***. In the implementation, there was no place that shows these two variables updating their value after initializing. Thus, how can these two variables be updated their values? Would you please also clarify what these two variables are used for?
```
 # dispatch BTPMessages
        for serialized_msg in serialized_msgs:
            try:
                msg = BTPMessage.from_bytes(serialized_msg)
            except BaseException as e:
                # TODO [TBD] ignore BTPMessage parse failure?
                Logger.warning(f"fail to parse BTPMessage err:{e}", TAG)
            else:
                if msg.dst == self._props.btp_addr:
                    self._handle_msg(prev, msg)
                else:
                    try:
                        next_link, dst = self._resolve_route(msg.dst.net)
                        self._send_message(next_link.addr, serialized_msg)
                    except BTPException as e:
                        self._send_error(prev, msg, e)
        link.rx_seq = link.rx_seq + len(serialized_msgs)
        if link.sack_term > 0 and link.sack_next <= self.block_height:
            self._send_sack(prev, status["height"], link.rx_seq)
            link.sack_next = link.sack_next + link.sack_term
```

Thanks for your time to reply my questions.

**Comment by jspark-icon** [comment, marked as answer]
SACKMessage is not related Relay-rotation, it is just same as heartbeat between connected BMC.
sack_height and sack_seq are representation of status of Link.
sack_term is setting value for SACK feature, sack_next is a threshold for sending.
SACK feature isn't yet determined to apply to BMC.

**Reply by quangtran8588** [reply]
Got it. Thanks for your clarification.
